clippy_lints/src/extra_unused_type_parameters.rs

```diff
@@ -0,0 +1,175 @@
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn no_unused_ty(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub EXTRA_UNUSED_TYPE_PARAMETERS,
+    complexity,
+    "unused type parameters in function definitions"
+}
+declare_lint_pass!(ExtraUnusedTypeParameters => [EXTRA_UNUSED_TYPE_PARAMETERS]);
+
+struct TypeWalker<'cx, 'tcx> {
+    cx: &'cx LateContext<'tcx>,
+    map: FxHashMap<DefId, Span>,
+    bound_map: FxHashMap<DefId, Span>,
+    generics: &'tcx Generics<'tcx>,
+    some_params_used: bool,
+    has_non_ty_params: bool,
+}
+
+impl<'cx, 'tcx> TypeWalker<'cx, 'tcx> {
+    fn new(cx: &'cx LateContext<'tcx>, generics: &'tcx Generics<'tcx>) -> Self {
+        let mut has_non_ty_params = false;
+        let map = generics
+            .params
+            .iter()
+            .filter_map(|param| match &param.kind {
+                GenericParamKind::Type { .. } => Some((param.def_id.into(), param.span)),
+                GenericParamKind::Lifetime {
+                    kind: LifetimeParamKind::Elided,
+                } => None,
+                _ => {
+                    has_non_ty_params = true;
+                    None
+                },
+            })
+            .collect();
+        Self {
+            cx,
+            map,
+            generics,
+            has_non_ty_params,
+            bound_map: FxHashMap::default(),
+            some_params_used: false,
+        }
+    }
+
+    fn emit_lint(&self) {
+        let (msg, help) = match self.map.len() {
+            0 => return,
+            1 => (
+                "type parameter goes unused in function definition",
+                "consider removing the parameter",
+            ),
+            _ => (
+                "type parameters go unused in function definition",
+                "consider removing the parameters",
+            ),
+        };
+
+        let source_map = self.cx.tcx.sess.source_map();
+        let span = if self.some_params_used || self.has_non_ty_params {
+            MultiSpan::from_spans(
+                self.map
+                    .iter()
+                    .map(|(def_id, &(mut span))| {
+                        if let Some(bound_span) = self.bound_map.get(def_id) {
+                            span = span.with_hi(bound_span.hi());
+                        }
+                        span = source_map.span_extend_to_next_char(span, ',', false);
+                        span = span.with_hi(BytePos(span.hi().0 + 1));
+
+                        let max_hi = self.generics.span.hi();
+                        if span.hi() >= max_hi {
+                            span = span.with_hi(BytePos(max_hi.0 - 1));
+                        }
+                        span
+                    })
+                    .collect(),
+            )
+        } else {
+            self.generics.span.into()
+        };
+
+        span_lint_and_help(self.cx, EXTRA_UNUSED_TYPE_PARAMETERS, span, msg, None, help);
```

Comment:
You've put so much effort in crafting this awesome spans and message, I think this should be `span_lint_and_sugg`. The non-multispan cases can then be tested with `// run-rustfix` in the test file. You'll have to split up the test files for that though: fixable and unfixable.